### PR TITLE
Add NPM launch type

### LINKED ISSUE #340: a NPM Launch Configuration
On a package.json, one should be able to do "Run As > npm update" (or whichever operation makes sense.

### PR DESCRIPTION
**Pull request by AObuchow**
Fix #340

Signed-off-by: Andrew Obuchowicz <aobuchow@redhat.com>

### CODE REVIEW

**Comment by AObuchow** [comment]
This probably is going to need tests. I can work on those monday.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/AbstractHTMLDebugDelegate.java [hunk 1]
```diff
@@ -55,12 +54,9 @@ public void launchWithParameters(ILaunchConfiguration configuration, String mode
 		} catch (Exception e) {
 			IStatus errorStatus = new Status(IStatus.ERROR, Activator.PLUGIN_ID, e.getMessage(), e);
 			Activator.getDefault().getLog().log(errorStatus);
-			Display.getDefault().asyncExec(new Runnable() {
-			@Override
-			public void run() {
-					ErrorDialog.openError(Display.getDefault().getActiveShell(), "Debug error", e.getMessage(), errorStatus); //$NON-NLS-1$
-			}
-			});
+			Display.getDefault().asyncExec(() -> ErrorDialog.openError(Display.getDefault().getActiveShell(),
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/AbstractHTMLDebugDelegate.java]: Unrelated cleanup (convert to Lambda)

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxRunDABDebugDelegate.java [hunk 2]
```diff
@@ -86,10 +86,11 @@ static File findDebugAdapter() {
 
 	private static String findFirefoxLocation() {
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxRunDABDebugDelegate.java]: Unrelated , but this code used to always return `"/path/to/firefox"`

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ * Copyright (c) 2018, 2019 Red Hat Inc. and others.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Mickael Istria (Red Hat Inc.) - initial implementation
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java]: Change contributors

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2019 Red Hat Inc. and others.
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java]: Update year

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/IsPackageJSONTester.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Red Hat Inc. and others.
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/IsPackageJSONTester.java]: Update year

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2019 Red Hat Inc. and others.
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java]: Update year

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Mickael Istria (Red Hat Inc.) - initial implementation
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java]: Update contributors

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/PackageJSONAdapterFactory.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Red Hat Inc. and others.
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/PackageJSONAdapterFactory.java]: Update year

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java (hunk 7 above)
Comment (mickaelistria) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java]: Or even get rid of the contributors area. It's not required, so we can live without it. it's less maintenance.

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 9]
```diff
@@ -38,6 +38,14 @@
             name="JSON"
             priority="normal">
       </content-type>
+      <content-type
+            base-type="org.eclipse.wildwebdeveloper.parent"
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: parent type should be `org.eclipse.wildwebdeveloper.json`

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 10]
```diff
@@ -38,6 +38,14 @@
...
+            file-extensions=".json"
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: Don't match file extension or it will match all json files

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 11]
```diff
@@ -38,6 +38,14 @@
...
+            file-names="package.json"
+            id="org.eclipse.wildwebdeveloper.packageJSON"
+            name="PackageJSON"
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: Let's name it "NPM package.json"

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 12]
```diff
@@ -38,6 +38,14 @@
             name="JSON"
             priority="normal">
       </content-type>
+      <content-type
+            base-type="org.eclipse.wildwebdeveloper.parent"
+            file-extensions=".json"
+            file-names="package.json"
+            id="org.eclipse.wildwebdeveloper.packageJSON"
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: Maybe `org.eclipse.wildwebdeveloper.json.npmpackage

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 13]
```diff
@@ -617,6 +663,13 @@
                type="org.eclipse.debug.ui.actions.ILaunchable">
          </adapter>
       </factory>
+      <factory
+            adaptableType="org.eclipse.core.resources.IResource"
+            class="org.eclipse.wildwebdeveloper.launch.npm.PackageJSONAdapterFactory">
+         <adapter
+               type="org.eclipse.wildwebdeveloper.Adapter1">
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: adapter1 ?

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/AbstractHTMLDebugDelegate.java [hunk 14]
```diff
@@ -32,14 +32,13 @@
 import org.eclipse.wildwebdeveloper.Activator;
 import org.eclipse.wildwebdeveloper.InitializeLaunchConfigurations;
 
-public class AbstractHTMLDebugDelegate extends DSPLaunchDelegate{
+public class AbstractHTMLDebugDelegate extends DSPLaunchDelegate {
 	public static final String PROGRAM = "program"; //$NON-NLS-1$
 	public static final String ARGUMENTS = "runtimeArgs"; //$NON-NLS-1$
-	public static final String CWD = DebugPlugin.ATTR_WORKING_DIRECTORY; //$NON-NLS-1$
+	public static final String CWD = DebugPlugin.ATTR_WORKING_DIRECTORY; // $NON-NLS-1$
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/AbstractHTMLDebugDelegate.java]: As there is no hardcoded stream on that line, `NON-NLS` should be removed.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java (hunk 7 above)
Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchShortcut.java]: Agreed, done.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+		String workingDirectory = this.packageJSONFile == null ? defaultSelectedFile.getParentFile().getAbsolutePath()
+				: this.packageJSONFile.getParentFile().getAbsolutePath();
+		configuration.setAttribute(AbstractHTMLDebugDelegate.CWD, workingDirectory);
+
+	}
+
+	@Override
+	public String getName() {
+		return "\u25B6\uFE0F Program";
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: Externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	@Override
+	public void initializeFrom(ILaunchConfiguration configuration) {
+		try {
+			defaultSelectedFile = getSelectedFile(shortcut::canLaunch);
+			String defaultSelectedFilePath = pathOrEmpty(defaultSelectedFile);
+			this.programPathText
+					.setText(configuration.getAttribute(AbstractHTMLDebugDelegate.PROGRAM, defaultSelectedFilePath));
+			this.argumentsCombo.setText(configuration.getAttribute(AbstractHTMLDebugDelegate.ARGUMENTS, "install"));
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: externalize `"install"`

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+			}
+			updateLaunchConfigurationDialog();
+		});
+
+		Button filePath = new Button(resComposite, SWT.PUSH);
+		filePath.setText("Browse...");
+		filePath.addSelectionListener(SelectionListener.widgetSelectedAdapter(e -> {
+			FileDialog filePathDialog = new FileDialog(resComposite.getShell());
+			filePathDialog.setFilterPath(getSelectedProject() == null ? null : getSelectedProject().getAbsolutePath());
+			filePathDialog.setText("Select a package.json to use");
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: Externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+		this.programPathText.addModifyListener(event -> {
+			setDirty(true);
+			File file = new File(programPathText.getText());
+			if (!file.isFile()) {
+				String errorMessage = Messages.RunProgramTab_error_unknownFile;
+				setErrorMessage(errorMessage);
+				decoration.setDescriptionText(errorMessage);
+				decoration.show();
+			} else if (!shortcut.canLaunch(file)) {
+				String errorMessage = "Not a package.json file";
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: Externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+			updateLaunchConfigurationDialog();
+		});
+
+		Composite emptySpace = new Composite(resComposite, SWT.NONE);
+		emptySpace.setLayoutData(new GridData(SWT.FILL, SWT.DEFAULT, false, false));
+
+		Composite fileComposite = new Composite(resComposite, SWT.NONE);
+		fileComposite.setLayout(new GridLayout(2, false));
+		fileComposite.setLayoutData(new GridData(SWT.FILL, SWT.DEFAULT, true, false));
+		new Label(fileComposite, SWT.NONE).setText("package.json file:");
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+
+	@Override
+	public void createControl(Composite parent) {
+		resComposite = new Composite(parent, SWT.NONE);
+		resComposite.setLayout(new GridLayout(2, false));
+
+		Composite argumentsComposite = new Composite(resComposite, SWT.NONE);
+		argumentsComposite.setLayout(new GridLayout(2, false));
+		argumentsComposite.setLayoutData(new GridData(SWT.FILL, SWT.DEFAULT, true, false));
+		new Label(argumentsComposite, SWT.NONE).setText("NPM command:");
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java (hunk 16 above)
Comment (mickaelistria) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: no need to externalize here, this is not a translatable text.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+
+		File packageJSONDirectory = new File(
+				configuration.getAttribute(AbstractHTMLDebugDelegate.CWD, "No package.json directory path set").trim());
+		File packageJSON = new File(
+				configuration.getAttribute(AbstractHTMLDebugDelegate.PROGRAM, "No package.json path set").trim());
+		String argumentString = configuration.getAttribute(AbstractHTMLDebugDelegate.ARGUMENTS, "No NPM argument set")
+				.trim();
+		List<String> arguments = new ArrayList<>();
+		arguments.add(findNPMLocation());
+		arguments.addAll(Arrays.asList(argumentString.split(" ")));
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java]: Maybe use a better regex in a future iteration? Something that will account for quotes (ie `npm run echo "hello world"`)

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+		monitor.beginTask("NPM install for: " + packageJSON.getAbsolutePath(), 2);
+		monitor.worked(1);
+		System.out.println("Dir" + packageJSONDirectory.getAbsolutePath());
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java]: remove println's (I've been using these cause Eclipse has been crashing sometimes when debugging launch configurations.. I'm not sure why. Maybe I need to upgrade my Eclipse install)

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+			try {
+				Process dependencyInstaller = new ProcessBuilder(arguments)
+						.directory(packageJSONDirectory).start();
+				dependencyInstaller.waitFor();
+				String operationOutput = null;
+				operationOutput = convertInputStreamToString(dependencyInstaller.getInputStream());
+				String errorOutput = null;
+				errorOutput = convertInputStreamToString(dependencyInstaller.getErrorStream());
+				if (!operationOutput.equals("")) {
+					Activator.getDefault().getLog().log(new Status(1, Activator.PLUGIN_ID, "NPM: " + operationOutput));
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java]: Externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+				}
+				if (!errorOutput.equals("")) {
+					Activator.getDefault().getLog().log(new Status(2, Activator.PLUGIN_ID, "NPM error: " + errorOutput));
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java]: Externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+				IProject resource = ResourcesPlugin.getWorkspace().getRoot().getProject(packageJSONDirectory.getName());
+				if (resource != null) {
+					resource.refreshLocal(IResource.DEPTH_INFINITE, new NullProgressMonitor());
+				}
+				monitor.worked(1);
+			} catch (InterruptedException | IOException | CoreException e) {
+				IStatus errorStatus = new Status(IStatus.ERROR, Activator.PLUGIN_ID, e.getMessage(), e);
+				Activator.getDefault().getLog().log(errorStatus);
+				Display.getDefault().asyncExec(() -> ErrorDialog.openError(Display.getDefault().getActiveShell(),
+						"NPM error", e.getMessage(), errorStatus)); //$NON-NLS-1$
```
Comment (AObuchow) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NpmLaunchDelegate.java]: Externalize

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java (hunk 16 above)
Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/launch/npm/NPMLaunchTab.java]: good call thanks

**Comment by AObuchow** [comment]
> I could see it working, the UI seems good enough.
> However, I see a few usability issues:
> 
> * absence of logging in the Console view

My latest patch addresses this. Hopefully it's done correctly.

**Comment by mickaelistria** [comment]
This got merged after a rebase.
